model_discretization_api.py

```diff
@@ -0,0 +1,209 @@
+
+
+# the convolution method can be set via string, to choose between the astropy.convolution
+# methods
+convolved_model = ConvolvedModel(source, psf, method='fft')
+convolved_model = ConvolvedModel(source, psf, method='standard')
+
+# or by passing a function that should be used. This is mainly nececessary, because
+# the performance of the astropy convolution algorithms is rather weak...
+convolved_model = ConvolvedModel(source, psf, method=scipy.signal.fftconvolve)
+
+# recommendation would be to use scipy.signal.fftconvolve which is very fast
+
+# additional keyword arguments can be passed to the convolution function
+convolved_model = ConvolvedModel(source, psf, method='standard', mode='oversample')
+
+# the convolved_model can be fit 
+
+# Further notes:
+# 1. Implementation of this class should be straight forward and not much effort
+# 2. Testing could be set up using Gaussian models, where the analytical solution is
+# known
+# 3. Later one could think about assigning the corresponding fourier transforms to analytical
+# models and use this for the convolution, which will probably have a better performance.
+# Source model and PSF are evaluated analytically in Fourier space, mutliplied and transformed
+# back to normal space.
+# 4. As convolution is a linear operation the derivative of a convolution behaves like 
+# http://en.wikipedia.org/wiki/Convolution#Differentiation this relation can be used
+# to implement `fit_deriv` methods for convolved models.
+
+
+#########################
+# Model integration API #
+#########################
+
+import numpy as np
+
+# to avoid bias, when the scale of the model is similar to the size of the pixels,
+# models have to be integrated over pixels, for this purpose models could define
+# an integrate function, that takes upper and lower bounds of the integration
+# (see also https://github.com/astropy/astropy/issues/1586)
+model = Gaussian1D(1, 0, 1)
+model.integrate(0, np.inf)
+
+# it should also work with arrays of bounds of course
+x_lo = np.arange(-10, 11) - 0.5
+x_hi = np.arange(-10, 11) + 0.5
+model.integrate((x_lo, x_hi))
+
+# for 2d models it would be
+y_lo = np.arange(-10, 11) - 0.5
+y_hi = np.arange(-10, 11) + 0.5
+model.integrate((x_lo, x_hi, y_lo, y_hi))
+
+# whenever possible (e.g. polynomials) the integrate function would be implemented
+# analytically, for all others a standard numerical integration routine would be used
+# e.g. from scipy.integrate
+# the integration mode can be passed to the integrate method if it is not already an 
+# attribute of model
+model.integrate((x_lo, x_hi, y_lo, y_hi), mode='integrate')
+
+# for convenience one could incorporate the functionality into the standard model API
+# as is done for the convolution kernals 
+model = Gaussian1D(1, 0, 1, mode='center') 
+model = Gaussian1D(1, 0, 1, mode='linear_interp') 
+model = Gaussian1D(1, 0, 1, mode='integrate')
+model = Gaussian1D(1, 0, 1, mode='oversample', factor=10)
```

Comment:
Could you explain a bit more about this?  What do "mode" and "factor" apply to when instantiating a model?